### ISSUE TITLE
[core] Insufficient deprecation warnings for XPath attributes

### ISSUE THREAD

**Issue by mmoehring** [pmd/pmd#2019]
**Affects PMD Version:** 6.18

**Rule:** Builtin rulesets 'Codestyle' and 'Design'

**Description:**
Starting with 6.18 I get deprecation warnings for the builtin rulesets codestyle and design.
I tracked down the two rulesets but there might be other builtin rulesets with warnings.

> Use of deprecated attribute 'VariableDeclaratorId/@Array' in XPath query
> Use of deprecated attribute 'VariableDeclaratorId/@ArrayDepth' in XPath query
> Use of deprecated attribute 'FieldDeclaration/@Array' in XPath query
> Use of deprecated attribute 'FieldDeclaration/@VariableName' in XPath query
> Use of deprecated attribute 'FieldDeclaration/@ArrayDepth' in XPath query
> Use of deprecated attribute 'PrimitiveType/@Array' in XPath query
> Use of deprecated attribute 'PrimitiveType/@ArrayDepth' in XPath query

The problem with this warning is that there is no indication of where it is coming from or which files are affected, which might be confusing. Examples: 
https://github.com/teamed/qulice/issues/1048
https://github.com/aistomin/maven-browser/issues/54

**Code Sample demonstrating the issue:**

Reference builtin rules codestyle or design in your ruleset:
```
<rule ref="category/java/codestyle.xml">
</rule>
...
<rule ref="category/java/design.xml">
</rule>
```

**Running PMD through:** *Gradle*

**Comment by oowekyala** [comment]
Yes, for now we don't track down attributes... The warnings themselves are wrong (at least those coming from built-in rulesets), the latest release had a bug (see e4ce2faf78)

@jsotuyod @adangel Maybe there's a way to fiddle with loggers to get a workaround, until we can fix that in 7.0.0? Something like a threadlocal with info about the currently executing rule maybe

**Comment by adangel** [comment]
> Something like a threadlocal with info about the currently executing rule maybe
That would be something for 7.0.0

But the triggering issue was, that we logged these deprecation warnings, when we shouldn't, right? This is fixed with e4ce2fa except for the case, when the attribute would return `null`. In order to fix that, we need to wrap the value or introduce a flag, so that we can distinguish the cases "method not called yet" and "method returned null".
I guess, we need to file a separate issue for that.
I'm wondering however, why we didn't see this in our builds - in https://github.com/pmd/pmd/blob/e4ce2faf78c2b08aa1663ebc1a76f110a09a9bba/pmd-java/src/test/java/net/sourceforge/pmd/coverage/PMDCoverageTest.java#L71 we explicitly check for these messages...

Fiddling with the logger, you potentially could disable the log entries of this class. One way is, to configure the global logging of the jvm (conf/logging.properties). Not sure, if that's worth it.

**Comment by oowekyala** [comment]
> I'm wondering however, why we didn't see this in our builds 

Yeah me too, though I spotted it when I couldn't build the 7.0.x branch locally...

> That would be something for 7.0.0

Why would that be? Ideally 7.0.0 would be smarter about that, which we can probably do after removing Jaxen.

**Comment by adangel** [comment]
The solution with threadlocal sounded like having some kind of execution context (see also #1371 ) were we would collect violations, missing classes, errors, etc. during the execution. That's why I've seen this as a 7.0.0 topic.

The current solution only logs it once per attribute (that's why it doesn't appear in PMDCoverageTest again). This means, we would only report the first rule, that uses it and not all rules. IMHO, this should then also be fixed, so that you see all usages of the deprecated attribute.

**Comment by jsotuyod** [comment]
> The solution with threadlocal sounded like having some kind of execution context (see also #1371 ) were we would collect violations, missing classes, errors, etc. during the execution. That's why I've seen this as a 7.0.0 topic.

Yes, and this context should be analysis-bound, not thread-bound; knowing the same JVM may be reused to run multiple analysis, maybe even in parallel (ie: using a Gradle daemon on a multi-module project, each being analyzed by PMD); so it's something we probably need to pass around, or at least be able to know which threads are working on a single analysis and when it starts / ends for housekeeping.

**Comment by adangel** [comment]
> be able to know which threads are working on a single analysis

yes, I think, that should be possible. With that, we could still implement the "context holder" as a thread local, but the first thing a thread/task needs to is, to set this thread local to the current analysis context. I assume, once our thread/task is starting, it won't switch to other analysis executions, so this should be safe.

**Comment by adangel** [comment]
I gave it a try, see #2057 - I think, it's worth for 6.19.0.

**Comment by oowekyala** [comment]
I put together a fix in https://github.com/oowekyala/pmd/tree/xpath-warnings
Sample warning:
```
WARNING: Use of deprecated attribute 'VariableDeclaratorId/@Image' by XPath rule 'AvoidEnumAsIdentifier' (in ruleset 'Error Prone'), please use @Name instead
```
Unlike #2057 it doesn't use more global state, each XPath rule just gets it own attribute usage recorder, and access to attributes is tracked by the saxon and jaxen adapters

I can open the pr once #2377 is merged
